File: CineMetrics/modules/story_card.py
# modules/story_card.py
import io
import requests
from PIL import Image, ImageDraw

def generate_rewind_story_card(month_year, persona, total_films, total_hours, top_director, top_genres, poster_url=None):
    """Renders a sleek 1080x1920 Instagram Story Card for the Monthly Rewind."""
    card = Image.new('RGB', (1080, 1920), color=(15, 23, 42))
    draw = ImageDraw.Draw(card)
    
    # Accent top banner
    draw.rectangle([(0, 0), (1080, 24)], fill="#38bdf8")
    
    # Header Branding
    draw.text((80, 140), f"CINEMETRICS // {month_year} REWIND", fill="#38bdf8")
    draw.text((80, 200), "Monthly Taste Capsule", fill="#f8fafc")
    
    # Persona Card
    draw.rounded_rectangle([(80, 320), (1000, 580)], radius=20, fill=(30, 41, 59), outline="#38bdf8", width=2)
    draw.text((120, 360), "YOUR CINEMATIC PERSONA", fill="#94a3b8")
    draw.text((120, 420), f"{persona}", fill="#f59e0b")
    
    # Metric 1: Hours & Films
    draw.rounded_rectangle([(80, 620), (520, 920)], radius=20, fill=(30, 41, 59))
    draw.text((120, 660), "WATCH TIME", fill="#38bdf8")
    draw.text((120, 720), f"{total_hours:.1f} hrs", fill="#ffffff")
    draw.text((120, 820), f"{total_films} films logged", fill="#94a3b8")
    
    # Metric 2: Top Director
    draw.rounded_rectangle([(560, 620), (1000, 920)], radius=20, fill=(30, 41, 59))
    draw.text((600, 660), "TOP AUTEUR", fill="#a855f7")
    draw.text((600, 720), f"{top_director[:18]}", fill="#ffffff")
    draw.text((600, 820), "Most watched director", fill="#94a3b8")

    # Metric 3: Top Genres
    draw.rounded_rectangle([(80, 960), (1000, 1260)], radius=20, fill=(30, 41, 59))
    draw.text((120, 1000), "FREQUENT VIBES & GENRES", fill="#10b981")
    genres_str = " • ".join(top_genres[:4]) if top_genres else "Cinema"
    draw.text((120, 1070), genres_str[:45], fill="#ffffff")

    # Optional Highlight Poster
    if poster_url:
        try:
            res = requests.get(poster_url, timeout=3)
            p_img = Image.open(io.BytesIO(res.content)).convert('RGB')
            p_img = p_img.resize((240, 360), Image.Resampling.LANCZOS)
            card.paste(p_img, (420, 1320))
        except Exception:
            pass

    # Footer
    draw.text((80, 1800), "Created with CineMetrics", fill="#64748b")
    
    buf = io.BytesIO()
    card.save(buf, format="PNG")
    return buf.getvalue()

File: src/services/storyCard.js
// src/services/storyCard.js

export async function generateStoryCardBlob({
  monthYear,
  persona,
  totalFilms,
  totalHours,
  topDirector,
  topGenres = [],
  posterUrl = null
}) {
  const canvas = document.createElement('canvas');
  canvas.width = 1080;
  canvas.height = 1920;
  const ctx = canvas.getContext('2d');

  const safeMonthYear = String(monthYear || 'Month').toUpperCase();
  const safeDirector = String(topDirector || 'Various Auteurs');
  const safePersona = String(persona || 'Cinephile');

  // Matte Black Background
  ctx.fillStyle = '#0a0a0a';
  ctx.fillRect(0, 0, 1080, 1920);

  // Soft Red Accent Top Bar
  ctx.fillStyle = '#e05260';
  ctx.fillRect(0, 0, 1080, 16);

  // Top Header Branding
  ctx.fillStyle = '#e05260';
  ctx.font = 'bold 32px "Plus Jakarta Sans", -apple-system, sans-serif';
  ctx.fillText(`CINEMETRICS // ${safeMonthYear} REWIND`, 80, 150);

  ctx.fillStyle = '#f2f2f2';
  ctx.font = 'bold 46px "Plus Jakarta Sans", -apple-system, sans-serif';
  ctx.fillText('Monthly Film Capsule', 80, 220);

  // Persona Card
  drawRoundedRect(ctx, 80, 310, 920, 250, 14, '#181818', '#2a2a2a', 2);
  ctx.fillStyle = '#999999';
  ctx.font = 'bold 22px "Plus Jakarta Sans", sans-serif';
  ctx.fillText('CINEMATIC PERSONA', 120, 370);

  ctx.fillStyle = '#f2f2f2';
  ctx.font = 'bold 44px "Plus Jakarta Sans", sans-serif';
  ctx.fillText(safePersona, 120, 460);

  // Watch Time Box
  drawRoundedRect(ctx, 80, 600, 440, 280, 14, '#181818', '#2a2a2a', 1);
  ctx.fillStyle = '#e05260';
  ctx.font = 'bold 22px "Plus Jakarta Sans", sans-serif';
  ctx.fillText('WATCH TIME', 120, 660);

  ctx.fillStyle = '#ffffff';
  ctx.font = 'bold 60px "Plus Jakarta Sans", sans-serif';
  ctx.fillText(`${totalHours.toFixed(1)} hrs`, 120, 750);

  ctx.fillStyle = '#777777';
  ctx.font = '24px "Plus Jakarta Sans", sans-serif';
  ctx.fillText(`${totalFilms} films logged`, 120, 820);

  // Top Director Box
  drawRoundedRect(ctx, 560, 600, 440, 280, 14, '#181818', '#2a2a2a', 1);
  ctx.fillStyle = '#eab308';
  ctx.font = 'bold 22px "Plus Jakarta Sans", sans-serif';
  ctx.fillText('TOP DIRECTOR', 600, 660);

  ctx.fillStyle = '#ffffff';
  ctx.font = 'bold 42px "Plus Jakarta Sans", sans-serif';
  ctx.fillText(safeDirector.slice(0, 16), 600, 750);

  ctx.fillStyle = '#777777';
  ctx.font = '24px "Plus Jakarta Sans", sans-serif';
  ctx.fillText('Most watched auteur', 600, 820);

  // Top Genres Box
  drawRoundedRect(ctx, 80, 920, 920, 260, 14, '#181818', '#2a2a2a', 1);
  ctx.fillStyle = '#999999';
  ctx.font = 'bold 22px "Plus Jakarta Sans", sans-serif';
  ctx.fillText('TOP GENRES', 120, 980);

  ctx.fillStyle = '#f2f2f2';
  ctx.font = 'bold 38px "Plus Jakarta Sans", sans-serif';
  const genreStr = topGenres.slice(0, 3).join(' • ') || 'Cinema';
  ctx.fillText(genreStr.slice(0, 40), 120, 1070);

  // Lead Poster
  if (posterUrl) {
    try {
      const img = await loadImage(posterUrl);
      ctx.save();
      roundRectPath(ctx, 420, 1240, 240, 360, 12);
      ctx.clip();
      ctx.drawImage(img, 420, 1240, 240, 360);
      ctx.restore();
    } catch {
      // ignore
    }
  }

  // Footer branding
  ctx.fillStyle = '#555555';
  ctx.font = 'bold 24px "Plus Jakarta Sans", sans-serif';
  ctx.fillText('Cinefy // Personal Film Engine', 80, 1800);

  return new Promise(resolve => {
    canvas.toBlob(blob => resolve(blob), 'image/png');
  });
}

function drawRoundedRect(ctx, x, y, width, height, radius, fill, stroke = null, lineWidth = 1) {
  ctx.save();
  ctx.beginPath();
  roundRectPath(ctx, x, y, width, height, radius);
  if (fill) {
    ctx.fillStyle = fill;
    ctx.fill();
  }
  if (stroke) {
    ctx.strokeStyle = stroke;
    ctx.lineWidth = lineWidth;
    ctx.stroke();
  }
  ctx.restore();
}

function roundRectPath(ctx, x, y, width, height, radius) {
  ctx.beginPath();
  ctx.moveTo(x + radius, y);
  ctx.lineTo(x + width - radius, y);
  ctx.quadraticCurveTo(x + width, y, x + width, y + radius);
  ctx.lineTo(x + width, y + height - radius);
  ctx.quadraticCurveTo(x + width, y + height, x + width - radius, y + height);
  ctx.lineTo(x + radius, y + height);
  ctx.quadraticCurveTo(x, y + height, x, y + height - radius);
  ctx.lineTo(x, y + radius);
  ctx.quadraticCurveTo(x, y, x + radius, y);
  ctx.closePath();
}

function loadImage(src) {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.crossOrigin = 'anonymous';
    img.onload = () => resolve(img);
    img.onerror = reject;
    img.src = src;
  });
}
